### PR TITLE
Feature/spark parent run facet execution

### PR DESCRIPTION
**Pull request by wjohnson**
### Problem

The Databricks Synapse connector breaks its execution into two separate jobs and execution ids. This leads to a disjointed set of run ids when observing lineage for Synapse as a destination.

Closes: #670

### Solution

This change introduces a `ParentRunFacetBuilder` and removes the ParentRunFacet from the `buildRun`.

> **Note:** All schema changes require discussion. Please [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) for context.

- [ ] Your change modifies the [core](https://github.com/OpenLineage/OpenLineage/blob/main/spec/OpenLineage.json) OpenLineage model
- [ ] Your change modifies one or more OpenLineage [facets](https://github.com/OpenLineage/OpenLineage/tree/main/spec/facets)

If you're contributing a new integration, please specify the scope of the integration and how/where it has been tested (e.g., Apache Spark integration supports `S3` and `GCS` filesystem operations, tested with AWS EMR).

### Checklist

- [ ] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [ ] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [ ] Your changes are accompanied by tests (_if relevant_)
- [ ] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [ ] You've updated the [`CHANGELOG.md`](https://github.com/OpenLineage/OpenLineage/blob/main/CHANGELOG.md) with details about your change under the "Unreleased" section (_if relevant, depending on the change, this may not be necessary_)
- [ ] You've versioned the core OpenLineage model or facets according to [SchemaVer](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/schemaver) (_if relevant_)

### CODE REVIEW

**Comment by wjohnson** [comment]
@collado-mike if you have a spare cycle, would you mind taking a look at this PR and helping me understand why `testSqlEventWithJobEventEmitsOnce` is failing?

In the PR so far, I've implemented the class you had mentioned in #707 and am just trying to pass the tests before trying to solve the Synapse problem.

However, I'm ending up with a NullPointerException in the `testSqlEventWithJobEventEmitsOnce` test.

It looks like it's failing on the SparkConfigUtils class but runs successfully on @mobuchowski previous commits to main.

```
  <testcase name="testSqlEventWithJobEventEmitsOnce()" classname="io.openlineage.spark.agent.OpenLineageSparkListenerTest" time="14.853">
    <failure message="java.lang.NullPointerException" type="java.lang.NullPointerException">java.lang.NullPointerException
	at io.openlineage.spark.agent.util.SparkConfUtils.findSparkConfigKey(SparkConfUtils.java:23)
	at io.openlineage.spark.agent.util.SparkConfUtils.findSparkConfigKey(SparkConfUtils.java:19)
	at io.openlineage.spark.agent.facets.builder.ParentRunFacetBuilder.buildParentFacet(ParentRunFacetBuilder.java:47)
	at io.openlineage.spark.agent.facets.builder.ParentRunFacetBuilder.lambda$build$1(ParentRunFacetBuilder.java:30)
	at java.util.Optional.ifPresent(Optional.java:159)
	at io.openlineage.spark.agent.facets.builder.ParentRunFacetBuilder.build(ParentRunFacetBuilder.java:28)
	at io.openlineage.spark.agent.facets.builder.ParentRunFacetBuilder.build(ParentRunFacetBuilder.java:17)
	at io.openlineage.spark.api.CustomFacetBuilder.accept(CustomFacetBuilder.java:36)
	at io.openlineage.spark.agent.lifecycle.OpenLineageRunEventBuilder.lambda$null$26(OpenLineageRunEventBuilder.java:461)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at io.openlineage.spark.agent.lifecycle.OpenLineageRunEventBuilder.lambda$buildRunFacets$27(OpenLineageRunEventBuilder.java:461)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	at io.openlineage.spark.agent.lifecycle.OpenLineageRunEventBuilder.buildRunFacets(OpenLineageRunEventBuilder.java:461)
	at io.openlineage.spark.agent.lifecycle.OpenLineageRunEventBuilder.populateRun(OpenLineageRunEventBuilder.java:274)
	at io.openlineage.spark.agent.lifecycle.OpenLineageRunEventBuilder.buildRun(OpenLineageRunEventBuilder.java:255)
	at io.openlineage.spark.agent.lifecycle.OpenLineageRunEventBuilder.buildRun(OpenLineageRunEventBuilder.java:216)
	at io.openlineage.spark.agent.lifecycle.SparkSQLExecutionContext.start(SparkSQLExecutionContext.java:61)
	at io.openlineage.spark.agent.OpenLineageSparkListenerTest.testSqlEventWithJobEventEmitsOnce(OpenLineageSparkListenerTest.java:91)
```

Thank you for any insights!

**Comment by mobuchowski** [comment]
>It looks like it's failing on the SparkConfigUtils class but runs successfully on @mobuchowski previous commits to main.

It fails on the code you've added: 

```
io.openlineage.spark.agent.facets.builder.ParentRunFacetBuilder.buildParentFacet(ParentRunFacetBuilder.java:47)
```

Those conf values are simply not set for tests.

**Comment by mobuchowski** [comment]
@wjohnson want to revisit this?

**Comment by wjohnson** [comment]
> @wjohnson want to revisit this?

Yes - It's been on my personal backlog for a while and will be hoping to contribute this either by the end of August😕🤞

**Comment by wjohnson** [comment]
I am super stoked to get this PR in and finally make the parent facet be part of the RunFacetBuilders rather than a bespoke component.  This lets us define more complex behaviors and helps to handle a unique situation with Databricks' Synapse / SQL Data Warehouse connector.

Thank you @hmoazam for helping me to unblock the initial issues with tests that @mobuchowski called out and thank you @collado-mike for the guidance on how this could be implemented!

File: integration/spark/shared/src/test/java/io/openlineage/spark/agent/util/PlanUtilsTest.java [hunk 1]
```diff
@@ -128,4 +131,18 @@ void testSchemaFacetWithListOfAttributes() {
...
+  @ParameterizedTest
+  @CsvSource({
+    "A Test Application,a_test_application",
+    "MyTestApplication,my_test_application",
+    "MyXMLBasedApplication,my_xml_based_application",
+    "JDBCRelationApplication,jdbc_relation_application",
+    "Test With a Single LetterBetweenWords,test_with_a_single_letter_between_words"
+  })
+  void testCamelCaseToSnakeCase(String appName, String expected) {
+    String actual =
```
[the remark this answers is not in the document]
Comment (wjohnson) [reply integration/spark/shared/src/test/java/io/openlineage/spark/agent/util/PlanUtilsTest.java:144]: In this case, I was just copying over an existing test which was testing the Regex I believe.  In addition, getSparkJobName needs the sparkContext 😞

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/PlanUtils.java [hunk 2]
```diff
@@ -295,4 +306,17 @@ public static <T, D> List<T> safeApply(PartialFunction<D, List<T>> pfn, D x) {
...
+  public static String getSparkJobName(SparkContext sparkContext, SparkPlan node) {
+    // Unwrap SparkPlan from WholeStageCodegen, as that's not a descriptive or helpful job name
+    if (node instanceof WholeStageCodegenExec) {
+      node = ((WholeStageCodegenExec) node).child();
+    }
+    return sparkContext
+            .appName()
+            .replaceAll(CAMEL_TO_SNAKE_CASE_REGEX, "_$1")
+            .toLowerCase(Locale.ROOT)
+        + "."
```
[the remark this answers is not in the document]
Comment (wjohnson) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/PlanUtils.java:319]: Excellent call!

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/ParentRunFacetBuilder.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+              parentFacet.ifPresent(facet -> consumer.accept("parent", facet));
+            });
+  }
+
+  private Optional<ParentRunFacet> buildParentFacet(SparkSession sess, SparkListenerEvent event) {
+    String parentRunId = null;
+    String parentJobName = null;
+
+    if (event instanceof SparkListenerSQLExecutionStart) {
+      // If this is an ExecutionStart, we need to extract the runId and jobName
```
[the remark this answers is not in the document]
Comment (wjohnson) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/ParentRunFacetBuilder.java:56]: Great point!  I will break this into methods for each section.

**Comment by mobuchowski** [comment]
Going to close this one over year of inactivity - if you still want to work on this sometime then reopen @wjohnson

**Pull request closed without merging**